Document the plan for fixing torn state compiler (#1)
jaredpar/sdk: documentation/general/torn-sdk.md

> CSC : warning CS9057: The analyzer assembly '..\dotnet\sdk\8.0.200\Sdks\Microsoft.NET.Sdk.Razor\source-
generators\Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.dll' references version '4.9.0.0' of the compiler, which is newer
than the currently running version '4.8.0.0'.

To address this issue we are going to separate Visual Studio and the .NET SDK in build scenarios. That will change our matrix to the following:

| Scenario | Loads Roslyn | Loads Analyzers / Generators |
| --- | --- | --- |
| msbuild | From .NET SDK | From .NET SDK |
| dotnet msbuild | From .NET SDK | From .NET SDK |
| Visual Studio Design Time | From Visual Studio | From Visual Studio |

Specifically we will be

1. Changing msbuild to use a compiler from the .NET SDK
2. Changing Visual Studio to use analyzers from Visual Studio

> In addition to making our builds more reliable this will also massively simplify our analyzer development strategy. Analyzers following this model can always target the latest Roslyn version without the need for complicated multi-targeting.

Reviewer: > Analyzers following this model can always target the latest Roslyn version

Do all in-box analyzers _need_ to target latest? Or is this a thing that we would like do to simplify the development/build/etc matrix? I'm wondering if there's some set of analyzers that could just target Ye Olde Roslyn Versions back enough to sidestep the issue.

Reviewer 2: This is a classic it depends; some analyzers won't care and could target Roslyn 1.0. But many will need to update their logic for newer language constructs and can't do that today, so they do things like this: https://github.com/dotnet/roslyn-analyzers/tree/main/src/Utilities/Compiler/Lightup.

Author: I will try and clarify a bit better about inbox analyzers. But essentially inbox analyzers have a few choices: 

1. Dual insert into VS: this is the recommended approach. 
2. Reference the oldest supported Roslyn version: this is just fine if it works for you. 
3. Multi-target like NuGet packages: this is not ideal because it increases .NET SDK size and it means that builds are significantly different between `msbuild` and `dotnet build`.

## Detailed Design

### MSBuild using .NET SDK Compiler

> The .NET SDK will start producing a package named Microsoft.NETSdk.Compiler.Roslyn. This will contain a .NET Framework version of the Roslyn compiler that matches .NET Core version. This will be published as a part of the .NET SDK release process meaning it's available for all publicly available .NET SDKs.

Author: I did the bad thing and named something: Microsoft.NETSdk.Compiler.Roslyn. 

Given my history of naming this will not last and a new name will be chosen. I don't really have any objections to a better name but ... whomever picks the new name is responsible for fixing up all the occurrences in this document :smile:

Reviewer 3: [Microsoft.NET.Sdk* packages](https://www.nuget.org/packages?q=microsoft.net.sdk) exist (with a dot between NET and Sdk unlike your proposed name) but they are used for dotnet workloads I think

Reviewer 2: Is the package specific to framework? IE, it will _only_ have the framework compilers (arm and x86)?

Author: Initially it will only have x64 framework compilers. Those will still run just fine an ARM64 via emulation. If there is significant ask we can later start shipping an ARM64 variant but I'd rather wait for that need to arise.

Author: Going to go with @jjonescz suggestion on the naming.

Reviewer 3: I wasn't suggesting a name, I just pointed out a similar name already exists. I'm not sure if we can reuse the same prefix for this package.

Reviewer 4: May I suggest Microsoft.Net.Sdk.Compilers.Toolset. That matches other sdk naming of other packages as mentioned above and also matches the naming of the compiler toolset package we already have. I don't see Roslyn used commonly in our package names.

Author: That is fine by me but note that when we generalize Roslyn as Compiler there is often some pushback from people that Roslyn is not the only .NET language. That is why I put Roslyn in the name here. But I'm guessing we're the only compiler that realistically has this type of problem given how the SDK worsk today.

The .NET SDK has the capability to [detect a torn state][pr-detect-torn-state]. When this is detected the matching Microsoft.NETSDK.Compiler.Roslyn package for this version of the .NET SDK will be downloaded via `<PackageDownload>`. Then the `$(RoslynTargetsPath)` will be reset to point into the package contents. This will cause MSBuild to load the Roslyn compiler from that location vs. the Visual Studio location.

> One downside to this approach is that it is possible to end up with two VBCSCompiler server processes. Consider that when a solution has a mix of .NET SDK style projects and legacy projects then in a torn state the .NET SDK projects will use the .NET SDK compiler will legacy projects will use the Visual Studio compiler. While not a desirable outcome, it is a correct one. Customers who wish to only have one VBCSCompiler should correct the torn state in their build tools.

Reviewer: Good call-out - the first thing that comes to mind is how this interacts with the changes we need to make for `dotnet build-server shutdown`. Would we be able to do any tracking here? Or I guess we'd still need the MSBuild feature to track and shutdown a long-running process...

Author: For `build-server-shutdown` there are still two paths in my mind: 

1. The ideal path: Can't remember if you or @rainersigwald filed the issue on how to do this better. Can't find the link at the moment. 
2. Best effort in current setup: the good news is that this still spawns process named VBCSCompiler so we can just kill that by name. 

For now we'd do (2).

Reviewer 5: Agree that status quo is ok here. I think you were hunting https://github.com/dotnet/msbuild/issues/10035.

Author: Added a section in work breakdown for the build-server shutdown work.

> One downside to this approach is that it is possible to end up with two VBCSCompiler server processes. Consider that when a solution has a mix of .NET SDK style projects and legacy projects then in a torn state the .NET SDK projects will use the .NET SDK compiler will legacy projects will use the Visual Studio compiler. While not a desirable outcome, it is a correct one. Customers who wish to only have one VBCSCompiler should correct the torn state in their build tools.

Reviewer 3: What does "legacy projects" mean here, like .NET Framework?

Author: These are the format of msbuild files before the .NET SDK was created. The easiest way to distinguish between them is the presence / absence of the `Sdk =` attribute in the `<Project>` element. It's a bit more nuanced than that but it's a quick test.

They build using the tasks / targets inside of MSBuild not the .NET SDK.

Reviewer 5: ```suggestion
One downside to this approach is that it is possible to end up with two VBCSCompiler server processes. Consider that when a solution has a mix of .NET SDK style projects and non-SDK .NET Framework projects then in a torn state the .NET SDK projects will use the .NET SDK compiler will legacy projects will use the Visual Studio compiler. While not a desirable outcome, it is a correct one. Customers who wish to only have one VBCSCompiler should correct the torn state in their build tools.
```

Author: Will add this in.

Author: Cleaned up the terminology a bit here too.

### Visual Studio using Visual Studio Analyzers

#### .NET SDK in box analyzers

Analyzers which ship in the .NET SDK box will change to having a copy checked into Visual Studio. When .NET SDK based projects are loaded at design time, the Visual Studio copy of the analyzer will be loaded. Roslyn already understands how to prefer Visual Studio copies of analyzers. That work will need to be extended a bit but that is pretty straight forward code.

This approach enables us to take actions like NGEN or R2R analyzers inside of Visual Studio. This is a long standing request from the Visual Studio perf team but very hard to satisfy in our current

This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out working well for this scenario. There is initial upfront work to get the in box copy loading but it has virtually zero maintanence cost.

> This also means these analyzers can vastly simplify their development model by alwasy targeting the latest version of Roslyn. There is no more need to multi-target beacuse the version of the compiler the analyzer will be used with is known at ship time for all core build scenarios.

Reviewer 2: ```suggestion
This also means these analyzers can vastly simplify their development model by always targeting the latest version of Roslyn. There is no more need to multi-target because the version of the compiler the analyzer will be used with is known at ship time for all core build scenarios.
```

This means that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagonstics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make. Customers who wish to have a consistent experience between design time should not operate in a torn state.

#### NuGet based analyzers

Analyzers which ship via NuGet will continue to following the existing [support policies][matrix-of-paine]. This means that they will either need to target a sufficiently old version of Roslyn or implement multi-targeting support.

## Work Breakdown

- [ ]: Flow the Micosoft.Net.Compilers.Toolset.Framework package to the .NET SDK
- [ ]: Create a new package Micosoft.NETSdk.Compiler.Roslyn in .NET SDK.
  - [ ]: The contents of this package will include the contents of the Microsoft.Net.Compilers.Toolset.Framework package

>   - [ ]: The contents will include a README.md stating the package is **not** supported for direct user consumption.

Reviewer: Should the package be unlisted after upload? That allows restores but not most downloads/direct use.

Author: Unlisting is a good idea. I'm also considering changing the package such that it's not usable via `<PackageReference>`. That would _significantly_ reduce the incentive to download it.

  - [ ]: The package will follow the versioning scheme of the .NET SDK.
- [ ]: Change the Sdk.targets file to have copies of the following three `<UsingTasks>` from [Microsoft.Common.tasks][microsoft-common-tasks]. Having a copy in Sdk.targets means that resetting `$(RoslynTargetsPath)` during build will change the chosen compiler.
- [ ]: When the .NET SDK detects a torn state
  - [ ]: Use a `<PackageDownload>` to acquire the Microsoft.NETSdk.Compiler.Roslyn package
  - [ ]: Change `$(RoslynTargetsPath)` to point into the package contents

## Alternative

### PackageReference

Instead of `<PackageDownload>` the toolset package could be installed via `<PackageReference>`. This is how the existing Microsoft.Net.Compilers.Toolset package works today. This has a number of disadvantages:

1. PackageReferences are not side effect free and have been shown to cause issues in complex builds. For example this package did not work by default in the Visual Studio or Bing builds. Deploying this at scale to all our customers will almost ceratinly cause issues.

> 2. This approach does not solve all scenarios. The mechanics of restore mean that the compiler is not swapped until part way through the build. Build operations that happen earlier, such as legacy WPF, will still use the Visual Studio compiler and hence experience analyzer compatibility issues.

Reviewer: > Build operations that happen earlier, such as legacy WPF, will still use the Visual Studio compiler and hence experience analyzer compatibility issues.

I'd love to learn more about this - WPF builds things before _Restore_?!

Author: @KirillOsenkov knows the details of why this occurs. I don't fully understand but I trust that he is correct on this.

Reviewer 5: I believe it's not that it builds _before_ restore, it's that (older copies of) the WPF task/targets _don't include restore outputs_ and thus aren't affected by restore. https://github.com/dotnet/wpf/issues/9112 has more details if you were feeling too cheerful today.

Reviewer 6: Yes, it's a mess. There can be three modes: 

 1. non-SDK (legacy) projects targeting desktop framework and using legacy WPF, 
 2. SDK-style projects targeting desktop framework and legacy WPF, and 
 3. SDK-style projects targeting modern .NET and modern open-source WPF (UseWPF=true). 

Only the third (modern) mode has the fixes to see PackageReferences and restore outputs. The first two import the targets from C:\Windows :(

1. Changes the process name from VBCSCompiler to `dotnet`. That seems insignificant but the process name is important as it's referenced in a lot of build scripts. This would be a subtle breaking change that we'd have to work with customers on.
2. This would make Visual Studio dependent on the servicing lifetime of the .NET Core runtime in the SDK. That has a very different lifetime and could lead to surprising outcomes down the road. For example if policy removed out of support runtimes from machines it would break the Visual Studio build.

## Related Issues

- https://github.com/dotnet/roslyn/issues/72672
- https://github.com/dotnet/installer/pull/19144

[microsoft-common-tasks]: https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Common.tasks#L106-L109
[matrix-of-paine]: https://aka.ms/dotnet/matrix-of-paine
[pr-detect-torn-state]: https://github.com/dotnet/installer/pull/19144
[code-razor-vs-load]: https://github.com/dotnet/roslyn/blob/9aea80927e3d4e5a2846efaa710438c0d8d2bfa2/src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs#L1009

## Work Breakdown

This will be moved to an issue when this PR is opened against dotnet/sdk proper:

> - [ ]: Flow the Micosoft.Net.Compilers.Toolset.Framework package to the .NET SDK

Reviewer 3: This step is already done, right?

Author: Yes. I kept it here on a hedge that we may have to change the setup of it / ship a new package. If the doc ends up in this current state then will delete this when I file the issue.

| Scenario | Loads Roslyn | Loads Analyzers / Generators |
| --- | --- | --- |
| msbuild | From Visual Studio | From .NET SDK |
| dotnet msbuild | From .NET SDK | From .NET SDK |
| Visual Studio Design Time | From Visual Studio | From Both |

Generally this mixing of components is fine because Visual Studio will install a .NET SDK that is functionally paired with it. That is the compiler and analyzers are the same hence mixing poses no real issue. For example 17.10 installs .NET SDK 8.0.3xx, 17.9 installs .NET SDK 8.0.2xx, etc ... However when the .NET SDK is not paired with the Visual Studio version,then compatibility issues can, and will, arise. Particularly when the .NET SDK is _newer_ than Visual Studio customers will end up with the following style of error:

> CSC : warning CS9057: The analyzer assembly '..\dotnet\sdk\8.0.200\Sdks\Microsoft.NET.Sdk.Razor\source-
generators\Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators.dll' references version '4.9.0.0' of the compiler, which is newer
than the currently running version '4.8.0.0'.

> To address this issue we are going to separate Visual Studio and the .NET SDK in build scenarios. That will change our matrix to the following:

Reviewer 4: You should include data to back up that this is a common problem. Here's the data I have from back in December:
![image](https://github.com/jaredpar/sdk/assets/12663534/ff83f0c9-15bb-41ec-aefc-59a1591c4f78)

Author: Does this telemetry include CI runs?

| Scenario | Loads Roslyn | Loads Analyzers / Generators |
| --- | --- | --- |
| msbuild | From .NET SDK | From .NET SDK |
| dotnet msbuild | From .NET SDK | From .NET SDK |
| Visual Studio Design Time | From Visual Studio | From Visual Studio |

Specifically we will be

1. Changing msbuild to use a compiler from the .NET SDK
2. Changing Visual Studio to use analyzers from Visual Studio

> In addition to making our builds more reliable this will also massively simplify our analyzer development strategy. Analyzers following this model can always target the latest Roslyn version without the need for complicated multi-targeting.

Reviewer 4: Include a link to the current POR: https://learn.microsoft.com/en-us/dotnet/core/porting/versioning-sdk-msbuild-vs#targeting-and-support-rules

Another reason is that teams use older Visual Studio versions due to internal constraints: like an organizational policy. At the same time they install the latest .NET SDK which puts them into a torn state.

This also hits any customer that uses a preview version of .NET SDK. These inherently represent a torn SDK state because they almost never match the compiler in Visual Studio This results in blockers for big teams like Bing from testing out our previews.

## Goals

This design has a number of goals:

1. The `msbuild` and `dotnet build` build experience should be equivalent.
1. The Visual Studio Design time experience is independent of the .NET SDK installed.
1. To make explicit that it is okay, and even expected, that the design time and command line build experiences can differ when the SDK is in a torn state.

## MSBuild using .NET SDK Compiler

The .NET SDK will start producing a package named Microsoft.NETSdk.Compiler.Roslyn. This will contain a .NET Framework version of the Roslyn compiler that matches .NET Core version. This will be published as a part of the .NET SDK release process meaning it's available for all publicly available .NET SDKs.

> The .NET SDK has the capability to [detect a torn state][pr-detect-torn-state]. When this is detected the matching Microsoft.NETSDK.Compiler.Roslyn package for this version of the .NET SDK will be downloaded via `<PackageDownload>`. Then the `$(RoslynTargetsPath)` will be reset to point into the package contents. This will cause MSBuild to load the Roslyn compiler from that location vs. the Visual Studio location.

Reviewer 4: do we need to worry about a malicious user overriding RoslynTargetsPath to their own evil version of the compiler?

Author: No. Any code which has the ability to override that property today could override the compiler in many other ways. For example you could change `CscToolPath` to change the compiler being used.

Author: Another way of thinking about it is that if malicious code has the ability to control MSBuild properties, there are many existing ways that they could hijack the compiler (or other tooling). The change here makes it more centralized but that just makes having a fully functional compiler easier, it doesn't change the attack surface

One downside to this approach is that it is possible to end up with two VBCSCompiler server processes. Consider that when a solution has a mix of .NET SDK style projects and legacy projects then in a torn state the .NET SDK projects will use the .NET SDK compiler will legacy projects will use the Visual Studio compiler. While not a desirable outcome, it is a correct one. Customers who wish to only have one VBCSCompiler should correct the torn state in their build tools.

## Visual Studio using Visual Studio Analyzers

### .NET SDK in box analyzers dual insert

Analyzers which ship in the .NET SDK box will change to having a copy checked into Visual Studio. When .NET SDK based projects are loaded at design time, the Visual Studio copy of the analyzer will be loaded. Roslyn already understands how to prefer Visual Studio copies of analyzers. That work will need to be extended a bit but that is pretty straight forward code.

> This approach enables us to take actions like NGEN or R2R analyzers inside of Visual Studio. This is a long standing request from the Visual Studio perf team but very hard to satisfy in our current

Reviewer 5: ```suggestion
This approach enables us to take actions like NGEN or R2R analyzers inside of Visual Studio. This is a long standing request from the Visual Studio perf team but very hard to satisfy with our current distribution strategy.
```

> This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out working well for this scenario. There is initial upfront work to get the Visual Studio copy loading in design time builds but it has virtually zero maintanence cost.

Reviewer 5: ```suggestion
This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out to work well for this scenario. There is initial upfront work to get the Visual Studio copy loading in design time builds but it has virtually zero maintenance cost.
```

> This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out working well for this scenario. There is initial upfront work to get the Visual Studio copy loading in design time builds but it has virtually zero maintanence cost.

Reviewer 4: You say virtually zero maintenance cost. Is that because Razor is already inserting into VS so including their generator was low marginal cost. What about teams like aspnet and windows desktop that don't have separate insertions today? Wouldn't this be adding additional insertion management to them separate from the SDK insertion?

Author: >  Is that because Razor is already inserting into VS so including their generator was low marginal cost

Basically yes. Once we had done the work insert the generator it just happens automatically as part of the Razor insertion. 

> What about teams like aspnet and windows desktop that don't have separate insertions today?

My expectation is that we'll settle on a design where .NET SDK insertions end up inserting all of the inbox analyzers. The copy used in the .NET SDK is the same that will be used in VS. Hence the most natural way to handle this is to insert these at the time the .NET SDK inserts into VS. That has the added bonus that we get VS perf runs on the experience as a whole.

This also means these analyzers can vastly simplify their development model by always targeting the latest version of Roslyn. There is no more need to multi-target because the version of the compiler the analyzer will be used with is known at ship time for all core build scenarios.

> This does mean that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagonstics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make for reliability. Customers who wish to have a consistent experience between design time should not operate in a torn state.

Reviewer 5: ```suggestion
This does mean that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagnostics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make for reliability. Customers who wish to have a consistent experience between IDE and build should not operate in a torn state.
```

> This does mean that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagonstics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make for reliability. Customers who wish to have a consistent experience between design time should not operate in a torn state.

Reviewer 4: Is there any way to detect this situation? My main concern with this approach is customer confusion where they see CLI warnings but not in VS or vice versa and end up thinking something is wrong with our tooling. No idea how to detect that and warn them since we shouldn't know they are doing a CLI build separately from VS.

Author: > Is there any way to detect this situation?

Going forward we should assume that if you're using .NET SDK style projects in a torn state then design time build and cli build will differ. That is essentially our stated policy. If the experiences are the same it's a coincidence, not a design.

## .NET SDK in box analyzers target oldest

Analyzers which have no need to target the latest Roslyn could instead choose to target the oldest supported version of Roslyn. That ensures they can safely load into any supported scenario without issue.

This strategy should be considered short term. The Roslyn API is under constant development to respond to the performance demands of Visual Studio. New APIs can quickly become virtually mandatory for analyzers and generators to use and will only be available on latest. At that point the analyzer will need to also update to use dual insertions.

### NuGet based analyzers

> Analyzers which ship via NuGet will continue to following the existing [support policies][matrix-of-paine]. This means that they will either need to target a sufficiently old version of Roslyn or implement multi-targeting support.

Reviewer 4: How many third party analyzers are there and do we have a forum for giving them a heads up on our changes here (and are they even aware of the matrix)?

Author: There are lots of 3rd party analyzers out there. Probably the best forum is announcements on the dotnet blog or announcement issues on the roslyn repo. 

At the same time, I'm not sure what we would announce there. Any NuGet based analyzer (3rd party or 1rst party) is subject to being run under multiple compiler versions as they can be installed in any .NET SDK. Analyzer multi-targeting is already done based on roslyn version, not SDK version. So that means there is really no change here at all.

Yes in a torn state they're likely to be run on a more recent compiler. That is the same compiler they'd be run on in a `dotnet  build` state so it's not a new scenario. 

I can overall expand on this a bit to make it clearer why this is not a meaningful change for them.

### PackageReference

Instead of `<PackageDownload>` the toolset package could be installed via `<PackageReference>`. This is how the existing Microsoft.Net.Compilers.Toolset package works today. This has a number of disadvantages:

1. PackageReferences are not side effect free and have been shown to cause issues in complex builds. For example this package did not work by default in the Visual Studio or Bing builds. Deploying this at scale to all our customers will almost ceratinly cause issues.
2. This approach does not solve all scenarios. The mechanics of restore mean that the compiler is not swapped until part way through the build. Build operations that happen earlier, such as legacy WPF, will still use the Visual Studio compiler and hence experience analyzer compatibility issues.
3. Mixing builds that do and do not restore can lead to different build outcomes as they can end up choosing different compilers. Mixing the version of msbuild / dotnet msbuild can also lead to strange outcomes as they can end up choosing different compilers.

### Install the Framework compiler

> An alternative is simply shipping both Roslyn compilers in the Windows .NET SDK. This means there is no need for a `<PackageDownload>` step as the compiler is always there. This is a simpler build but it means that all .NET SDK installs on Windows will incease by ~6-7%.

Reviewer 5: ```suggestion
An alternative is simply shipping both Roslyn compilers in the Windows .NET SDK. This means there is no need for a `<PackageDownload>` step as the compiler is always there. This is a simpler build but it means that all .NET SDK installs on Windows will increase by ~6-7%.
```

> An alternative is simply shipping both Roslyn compilers in the Windows .NET SDK. This means there is no need for a `<PackageDownload>` step as the compiler is always there. This is a simpler build but it means that all .NET SDK installs on Windows will incease by ~6-7%.

Reviewer 4: Having the windows and linux installs be significantly different can cause potential customer confusion (not sure if customers are comparing the two). Not sure if we can exclude this package from the container image per the conversation with Rich as that would be preferred in this configuration.

Author: Let me add a bit about these concerns here.

### Use the .NET Core compiler

> Instead of downloading a .NET Framework Roslyn in a torn state, the SDK could just use the .NET Core Roslyn it comes with. That would have virtually no real compatibility issues. This has a number of disadvantages:

Reviewer 5: ```suggestion
Instead of downloading a .NET Framework Roslyn in a torn state, the SDK could use the .NET Core Roslyn it comes with. That would have virtually no real compatibility issues. This has a number of disadvantages:
```

> 1. Changes the process name from VBCSCompiler to `dotnet`. That seems insignificant but the process name is important as it's referenced in a lot of build scripts. This would be a subtle breaking change that we'd have to work with customers on.

Reviewer 5: This is real but worth fixing IMO.

Author: Agree. Once we are in the VMR I think it's realistic that we can address this. It's really hard in the current product construction state though. 

Do we have an issue tracking this that I can link to here?

> 2. This would make Visual Studio dependent on the servicing lifetime of the .NET Core runtime in the SDK. That has a very different lifetime and could lead to surprising outcomes down the road. For example if policy removed out of support runtimes from machines it would break the Visual Studio build.

Reviewer 5: Yeah this is the harder part. How much worse is it than SDK lifecycle though? Is it even possible for the policy to remove stale runtimes that were installed as part of the SDK while leaving the SDK?

(Don't mean to push on this, though, I agree it's probably not the right direction for now)

Author: > Is it even possible for the policy to remove stale runtimes that were installed as part of the SDK while leaving the SDK?

It is my understanding that this is exactly what MS IT does today. Brought that up in several meetings and have not been challenged on it. Would _love_ to be wrong but do not think I am 😦

[matrix-of-paine]: https://aka.ms/dotnet/matrix-of-paine
[pr-detect-torn-state]: https://github.com/dotnet/installer/pull/19144
[code-razor-vs-load]: https://github.com/dotnet/roslyn/blob/9aea80927e3d4e5a2846efaa710438c0d8d2bfa2/src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs#L1009
[setup-dotnet]: https://github.com/actions/setup-dotnet

## Work Breakdown

This will be moved to an issue when this PR is opened against dotnet/sdk proper:

- [ ]: Flow the Micosoft.Net.Compilers.Toolset.Framework package to the .NET SDK
- [ ]: Create a new package Micosoft.NETSdk.Compiler.Roslyn in .NET SDK.
  - [ ]: The contents of this package will include the contents of the `tasks\net472` folder in the Microsoft.Net.Compilers.Toolset.Framework package. This subset is all that is needed and makes the package not usable via `<PackageReference>`. The latter reduces the incentive for customers to use it directly.
  - [ ]: The contents will include a README.md stating the package is **not** supported for direct user consumption.
  - [ ]: The package will follow the versioning scheme of the .NET SDK.
  - [ ]: The package will be unlisted (ideal but not a hard requirement)

> - [ ]: Change the Sdk.targets file to have copies of the following three `<UsingTasks>` from [Microsoft.Common.tasks][microsoft-common-tasks]. Having a copy in Sdk.targets means that resetting `$(RoslynTargetsPath)` during build will change the chosen compiler.

Reviewer 5: 👍🏻

One downside to this approach is that it is possible to end up with two VBCSCompiler server processes. Consider a solution that has a mix of .NET SDK style projects and non-SDK .NET projects. In a torn the .NET SDK projects will use the .NET SDK compiler and non-SDK .NET projects will use the Visual Studio compiler. While not a desirable outcome, it is a correct one. Customers who wish to only have one VBCSCompiler should correct the torn state in their build tools.

## Visual Studio using Visual Studio Analyzers

### .NET SDK in box analyzers dual insert

Analyzers which ship in the .NET SDK box will change to having a copy checked into Visual Studio. When .NET SDK based projects are loaded at design time, the Visual Studio copy of the analyzer will be loaded. Roslyn already understands how to prefer Visual Studio copies of analyzers. That work will need to be extended a bit but that is pretty straight forward code.

This approach enables us to take actions like NGEN or R2R analyzers inside of Visual Studio. This is a long standing request from the Visual Studio perf team but very hard to satisfy in our current

> This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out working well for this scenario. There is initial upfront work to get the Visual Studio copy loading in design time builds but it has virtually zero maintanence cost.

Reviewer 3: ```suggestion
This approach is already taken by [the Razor generator][code-razor-vs-load]. This work was done for other reliability reasons but turned out working well for this scenario. There is initial upfront work to get the Visual Studio copy loading in design time builds but it has virtually zero maintenance cost.
```

This also means these analyzers can vastly simplify their development model by always targeting the latest version of Roslyn. There is no more need to multi-target because the version of the compiler the analyzer will be used with is known at ship time for all core build scenarios.

> This does mean that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagonstics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make for reliability. Customers who wish to have a consistent experience between design time should not operate in a torn state.

Reviewer 3: ```suggestion
This does mean that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagnostics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make for reliability. Customers who wish to have a consistent experience between design time should not operate in a torn state.
```

In the case of [multi-targeting][issue-analyzer-mt] this proposal is effectively a no-op. Analyzer multi-targeting is already based off of Roslyn versions, not .NET SDK versions, hence the proper version of the analyzer will still load in a torn state. The version of the Roslyn compiler in a torn state for msbuild is the same used for dotnet build hence it's already a scenario the analyzer needs to support.

## Alternative

### .NET SDK in box analyzers multi-target

Technically in box analyzers can have a multi-targeting strategy just as NuGet based analyzers do. This is actively discouraged because it leads to unnecessary increases in .NET SDK sizes. The time spent implementing multi-targeting is likely better spent moving to a dual insertion to keep .NET SDK sizes down and provide a consistent experience with other analyzers.

### PackageReference

Instead of `<PackageDownload>` the toolset package could be installed via `<PackageReference>`. This is how the existing Microsoft.Net.Compilers.Toolset package works today. This has a number of disadvantages:

> 1. PackageReferences are not side effect free and have been shown to cause issues in complex builds. For example this package did not work by default in the Visual Studio or Bing builds. Deploying this at scale to all our customers will almost ceratinly cause issues.

Reviewer 3: ```suggestion
1. PackageReferences are not side effect free and have been shown to cause issues in complex builds. For example this package did not work by default in the Visual Studio or Bing builds. Deploying this at scale to all our customers will almost certainly cause issues.
```

2. This approach does not solve all scenarios. The mechanics of restore mean that the compiler is not swapped until part way through the build. Build operations that happen earlier, such as [legacy WPF][issue-legacy-wpf], will still use the Visual Studio compiler and hence experience analyzer compatibility issues.
3. Mixing builds that do and do not restore can lead to different build outcomes as they can end up choosing different compilers. Mixing the version of msbuild / dotnet msbuild can also lead to strange outcomes as they can end up choosing different compilers.

### Install the Framework compiler

> An alternative is simply shipping both Roslyn compilers in the Windows .NET SDK. This means there is no need for a `<PackageDownload>` step as the compiler is always there. This is a simpler build but it means that all .NET SDK installs on Windows will incease by ~6-7%.

Reviewer 3: ```suggestion
An alternative is simply shipping both Roslyn compilers in the Windows .NET SDK. This means there is no need for a `<PackageDownload>` step as the compiler is always there. This is a simpler build but it means that all .NET SDK installs on Windows will increase by ~6-7%.
```

## Motivations

There are a number of scenarios where customers end up in a torn state. The most common is when customers have a CI setup with the following items:

1. They use a task like [actions/setup-dotnet][setup-dotnet] to install the .NET SDK and use a flexible version format like `8.x`.
2. They use `msbuild` to actually drive their build. This `msbuild` comes from the Visual Studio installed on the machine image.

This means that CI systems are updated to the latest .NET SDK virtually as soon as we release them. However the version of Visual Studio is updated much later as CI images usually take several weeks to upgrade to a new Visual Studio version. This is a very common CI setup and means a significant number of our customers end up in a torn state for several weeks.

Another reason is that teams use older Visual Studio versions due to internal constraints: like an organizational policy. At the same time they install the latest .NET SDK which puts them into a torn state.

> This also hits any customer that uses a preview version of .NET SDK. These inherently represent a torn SDK state because they almost never match the compiler in Visual Studio This results in blockers for big teams like Bing from testing out our previews.

Reviewer 3: ```suggestion
This also hits any customer that uses a preview version of .NET SDK. These inherently represent a torn SDK state because they almost never match the compiler in Visual Studio. This results in blockers for big teams like Bing from testing out our previews.
```
